text/vga-console.md

```diff
@@ -0,0 +1,223 @@
+# VGA Console
+
+The VGA spec and the guaranteed presence of VGA hardware on any machine or VM
+intermezzOS targets provide convenient methods of printing information for us to
+observe the kernel in action.
+
+## Details
+
+The VGA spec says that the magic address `0xB8000` is the start of video memory,
+and represents the top left cell on the screen. The screen is 80 columns wide,
+25 rows high, and each column is two bytes. In total, VGA video memory is 4000
+bytes (just under 4K).
+
+### Overview
+
+At present, intermezzOS' VGA implementation consists of a struct that owns a
+VGA-sized buffer **not** at the magic location, the VGA memory that is at the
+magic location, and some minor internal state.
+
+The console structure uses the buffer to collect text until `flush()` is called,
+at which point the buffer is copied en masse to the video memory. This provides
+apparently-atomic updates to the screen, although this operation need not be
+*actually* atomic as (a) good design and Rust checks should guarantee only one
+VGA structure owns video memory, and (b) moving 4k of memory is a long time to
+not have any other interrupts.
+
+### Proposed Design
+
+Philipp Oppermann suggests using a Unique pointer to wrap the magic memory
+location so that the compiler can guarantee that there is ever only one mutable
+owner of the video memory. I have not yet found a working way to magically
+assure Rust that the magic raw memory is the type we want -- I remember building
+this out of `core::ptr::from_raw_parts_mut()` in the workshop but have not been
+able to get it to work properly with any type other than `[u8; 4000]`.
+Regardless of *how* a VGA-memory-owner gets its memory, though, we will have to
+do magic to assure Rust that the raw memory at the VGA magic address is able to
+be used within the type system.
+
+I propose designing a VGA structure which owns only a single VGA-memory-sized
+slice, rather than the current implementation which gives one VGA structure two
+regions of memory. Since the magic address is the kind of implementation detail
+which should be hidden from consumers, the VGA structure will implement multiple
+constructors: the standard `new()` method will take a generic buffer, which can
+be allocated anywhere in normal memory (but as intermezzOS does not yet have a
+dynamic allocator, it will probably be in BSS), and an **unsafe** `new_at_vga()`
```

Comment:
Actually, instead of a constructor that owns the magic memory, we might be better off with an `unsafe fn Vga::display(&mut self)` function which briefly obtains mutable access to magic memory, copies `self` into it as bytes, and releases access. A kernel-scope Mutex can be used to guard access to magic memory (or have that be the only usage of the `new_at_vga()` ctor?) so that in the future if we get threads, there won't be a race to the screen.